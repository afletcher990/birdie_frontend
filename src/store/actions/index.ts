export * from './careRecipientActions';
export * from './dateActions';