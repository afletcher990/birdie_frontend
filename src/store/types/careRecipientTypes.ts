import { CareRecipient } from './careRecipientTypes';

export const SET_CARE_RECIPIENTS = 'SET_CARE_RECIPIENTS';
export const SET_CURRENT_CARE_RECIPIENT = 'SET_CURRENT_CARE_RECIPIENT';

export interface CareRecipient {
    care_recipient_id: string;
    name: string;
}

export interface CareRecipientInfo {
    careRecipients: Array<CareRecipient>;
    currentCareRecipient: CareRecipient;
}

interface SetCurrentCareRecipientAction {
    type: typeof SET_CURRENT_CARE_RECIPIENT;
    careRecipient: CareRecipient;
}

interface SetCareRecipientsAction {
    type: typeof SET_CARE_RECIPIENTS;
    careRecipientInfo: CareRecipientInfo;
}

export type CareRecipientActionTypes = SetCareRecipientsAction | SetCurrentCareRecipientAction;