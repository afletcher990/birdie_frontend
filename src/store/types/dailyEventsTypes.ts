export const SET_DAILY_EVENTS = 'SET_DAILY_EVENTS';

export interface CareEvent {
    visit_id: string;
    caregiver_id: string;
    caregiver_name: string;
    timestamp: Date;
    event_type: string;
    payload: {
        note: string;
        mood: string;
        task_definition_description: string;
    };
}

interface SetCareEventsAction {
    type: typeof SET_DAILY_EVENTS;
    payload: Array<CareEvent>;
}

export type CareEventActionTypes = SetCareEventsAction;