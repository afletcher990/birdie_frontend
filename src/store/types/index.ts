export * from './careRecipientTypes';
export * from './currentCareRecipientTypes';
export * from './viewDateTypes';
export * from './dailyEventsTypes';