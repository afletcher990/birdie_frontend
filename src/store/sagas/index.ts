import { put, call, all, select, takeEvery } from 'redux-saga/effects';
import { config } from './Constants';
import { history } from '@App/store';
import {
  SET_CARE_RECIPIENTS,
  SET_CURRENT_CARE_RECIPIENT,
  SET_DATE_INFO,
  SET_DAILY_EVENTS,
  CareRecipient, SET_VIEW_DATE
} from '@App/store/types';
import moment from 'moment';
import queryString from 'query-string';

moment.defaultFormat = 'YYYY-MM-DD';

function* initSaga() {
  yield all([
      getCareRecipients(),
      getValidDates(),
      watchDateChangeAsync()
  ]);
}

export default initSaga;

export function* watchDateChangeAsync() {
  yield takeEvery(SET_VIEW_DATE, getDailyEvents);
}

export function* getDailyEvents() {
  const state = yield select();
  const date = moment(state.dateInfo.viewDate).format();
  const careRecipientId = state.currentCareRecipient.care_recipient_id;
  const response = yield call(fetch, config.url.API_URL +
      `/getDailyEvents/?date=${date}&careRecipientId=${careRecipientId}`);
  const events = yield call([response, response.json]);
  yield put({type: SET_DAILY_EVENTS, payload: events});
}

export function* getValidDates() {
  const response = yield call(fetch, config.url.API_URL + '/getDatesWithVisits');
  const dateList = yield call([response, response.json]);
  const dates = dateList.map((item: {date: string}) => {
    return new Date(item.date);
  });
  const params = queryString.parse(history.location.search);
  let date: Date = dates[dates.length - 1];
  let found: boolean = false;
  if (params && params.date) {
    const findDate = dates.filter((item: Date) => {
      return (moment(item).format() === params.date);
    });
    if (findDate.length) {
      found = true;
      date = findDate[0];
    }
  }
  if (!found) {
    params.date = moment(date).format();
    history.push({
      ...history.location,
      search: `?${queryString.stringify(params)}`
    });
  }
  yield put({type: SET_DATE_INFO, payload: {
    validDates: dates, viewDate: date
    }});
}

export function* getCareRecipients() {
  const response = yield call(fetch, config.url.API_URL + '/getCareRecipients');
  const recipients = yield call([response, response.json]);
  yield put({type: SET_CARE_RECIPIENTS, payload: recipients});
  const params = queryString.parse(history.location.search);
  let recipient: CareRecipient = recipients[0];
  let found: boolean = false;
  if (params && params.care_recipient_id) {
    const findRecipient = recipients.filter((item: CareRecipient) => {
      return (item.care_recipient_id === params.care_recipient_id);
    });
    if (findRecipient.length) {
      found = true;
      recipient = findRecipient[0];
      params.care_recipient_id = recipient.care_recipient_id;
    }
  }
  if (!found) {
    params.care_recipient_id = recipients[0].care_recipient_id;
    history.push({
      ...history.location,
      search: `?${queryString.stringify(params)}`
    });
  }
  yield put({type: SET_CURRENT_CARE_RECIPIENT, payload: recipient});
}
