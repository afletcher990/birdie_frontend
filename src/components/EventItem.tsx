import styled from 'styled-components';
import * as React from 'react';
import moment from 'moment';
import { CareEvent } from '@App/store/types';

interface EventItemProps {
    key: string;
    item: CareEvent;
}

const getDisplayName = (id: string) => {
    return id.charAt(0).toUpperCase() + id.substring(1).replace(/_/g, ' ');
};

const EventItemContainer = styled.div`
    text-align: center;
    font-family: sans-serif;
    font-size: 16px;
    flex-direction: row;
    align-items: flex-start;
    display: flex;
    justify-content: center;
    margin: 0;
`;

const TimeBox = styled.div`
    border: grey;
    border-style: solid;
    margin: 6px;
    padding: 6px;
    border-radius: 10px;
    border-width: thin;
`;

const EventBox = styled.div`
    border: grey;
    border-style: solid;
    margin: 6px;
    padding: 6px;
    width: 20rem;
    border-radius: 10px;
    border-width: thin;
    text-align: start;
    
    @media(min-width: 768px) {
        width: 40rem;
    }
`;

const getEventDescription = (careEvent: CareEvent) => {
    switch (careEvent.event_type) {
        case 'check_in':
        case 'check_out':
        case 'visit_completed':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.caregiver_name})`;
        case 'task_completed':
            return careEvent.payload.task_definition_description;
        case 'mood_observation':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.mood})`;
        case 'general_observation':
            return careEvent.payload.note;
        case 'general_observation':
            return careEvent.payload.note;
        case 'fluid_intake_observation':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.fluid} 
${careEvent.payload.consumed_volume_ml}ml - ${careEvent.payload.observed ? 'Observed' : 'Not Observed'})`;
        case 'food_intake_observation':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.meal} - ${careEvent.payload.note})`;
        case 'incontinence_pad_observation':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.pad_condition})`;
        case 'catheter_observation':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.volume_ml}ml - 
${careEvent.payload.note})`;
        case 'mental_health_observation':
        case 'physical_health_observation':
        case 'regular_medication_not_taken':
            return `${getDisplayName(careEvent.event_type)} (${careEvent.payload.note})`;
        case 'alert_raised':
        case 'alert_qualified':
        case 'medication_schedule_updated':
        case 'regular_medication_maybe_taken':
        case 'medication_schedule_created':
        case 'regular_medication_partially_taken':
        case 'toilet_visit_recorded':
        case 'no_medication_observation_received':
        case 'regular_medication_taken':
        default:
            return getDisplayName(careEvent.event_type);
    }
    return '';
};

const getStyle = (eventType: string) => {
    switch (eventType) {
        case 'check_in':
            return {backgroundColor: 'lightGreen'};
        case 'check_out':
        case 'visit_completed':
            return {backgroundColor: '#f5aeb3'};
        case 'fluid_intake_observation':
        case 'food_intake_observation':
            return {backgroundColor: '#aedaf5'};
        case 'incontinence_pad_observation':
        case 'catheter_observation':
        case 'toilet_visit_recorded':
            return {backgroundColor: '#f5edae'};
        case 'regular_medication_maybe_taken':
        case 'medication_schedule_created':
        case 'regular_medication_partially_taken':
        case 'no_medication_observation_received':
        case 'regular_medication_taken':
            return {backgroundColor: '#f7b092'};
        default: return {};
    }
};

const EventItem = (props: EventItemProps) => {
    return (
        <EventItemContainer>
            <TimeBox>{moment(props.item.timestamp).format('HH:mm')}</TimeBox>
            <EventBox style={getStyle(props.item.event_type)}>{getEventDescription(props.item)}</EventBox>
        </EventItemContainer>
    );
};

export default EventItem;
