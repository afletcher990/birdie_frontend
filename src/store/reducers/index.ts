import { combineReducers } from 'redux';
import { currentCareRecipientReducer } from './currentCareRecipientReducer';
import { CareEvent, CareRecipient, DateInfo } from '@App/store/types';
import { viewDateReducer } from '@App/store/reducers/viewDateReducer';
import { careRecipientsReducer } from '@App/store/reducers/eventListReducer';

export type RootState = Readonly<{
    currentCareRecipient: CareRecipient,
    dateInfo: DateInfo,
    careEventList: Array<CareEvent>
}>;

export const rootReducer = combineReducers<RootState>({
    currentCareRecipient: currentCareRecipientReducer,
    dateInfo: viewDateReducer,
    careEventList: careRecipientsReducer
});
