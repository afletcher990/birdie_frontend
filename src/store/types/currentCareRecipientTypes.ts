import { CareRecipient } from './careRecipientTypes';

export const SET_CURRENT_CARE_RECIPIENT = 'SET_CURRENT_CARE_RECIPIENT';

interface SetCurrentCareRecipientAction {
    type: typeof SET_CURRENT_CARE_RECIPIENT;
    payload: CareRecipient;
}

export type CurrentCareRecipientActionTypes = SetCurrentCareRecipientAction;