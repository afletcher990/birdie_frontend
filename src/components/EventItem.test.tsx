import * as React from 'react';
import * as ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import EventItem from './EventItem';
import { CareEvent } from '@App/store/types';
const careEvent: CareEvent = {
    caregiver_id: '5c9090ab-7d5e-4a72-8bf7-197190ad4c98',
    caregiver_name: 'Skyla Cresswell',
    event_type: 'check_in',
    payload: {
        volume_ml: 0,
        pad_condition: '',
        meal: '',
        observed: false,
        consumed_volume_ml: 0,
        fluid: '',
        note: '',
        mood: '',
        task_definition_description: ''
    },
    timestamp: new Date('2019-05-12T06:34:30.825Z'),
    visit_id: '5cd753f0-8b66-f8a8-4591-3f78ca3f9c45'
};

const careEvent2: CareEvent = {
    caregiver_id: '3cacba0a-6041-11e9-b63f-06a80bfbb33e',
    caregiver_name: 'Rahim Kavanagh',
    event_type: 'fluid_intake_observation',
    payload: {
        volume_ml: 250,
        pad_condition: '',
        meal: '',
        observed: false,
        consumed_volume_ml: 0,
        fluid: 'regular',
        note: '',
        mood: '',
        task_definition_description: ''
    },
    timestamp: new Date('2019-05-12T07:21:11.990Z'),
    visit_id: '5cd753f0-8b66-f8a8-4591-3f78ca3f9c45'
};

it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<EventItem key="key1" item={careEvent}/>,  div);
});

it('Displays correct details', () => {
    const component = renderer.create(
        <EventItem key="key1" item={careEvent}/>,);
    let eventItem = component.toJSON();
    expect(eventItem).toMatchSnapshot();
});

it('Displays fluid intake correct details', () => {
    const component = renderer.create(
        <EventItem key="key1" item={careEvent2}/>,);
    let eventItem = component.toJSON();
    expect(eventItem).toMatchSnapshot();
});
