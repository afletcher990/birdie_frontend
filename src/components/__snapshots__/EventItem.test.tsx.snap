// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Displays correct details 1`] = `
<div
  className="sc-bdVaJa uKaCG"
>
  <div
    className="sc-bwzfXH fYfKuy"
  >
    07:34
  </div>
  <div
    className="sc-htpNat jLzanf"
    style={
      Object {
        "backgroundColor": "lightGreen",
      }
    }
  >
    Check in (Skyla Cresswell)
  </div>
</div>
`;

exports[`Displays fluid intake correct details 1`] = `
<div
  className="sc-bdVaJa uKaCG"
>
  <div
    className="sc-bwzfXH fYfKuy"
  >
    08:21
  </div>
  <div
    className="sc-htpNat jLzanf"
    style={
      Object {
        "backgroundColor": "#aedaf5",
      }
    }
  >
    Fluid intake observation (regular 
    0ml - Not Observed)
  </div>
</div>
`;
